File: 234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        tort, hare = head, head
        prev = None
        while hare is not None and hare.next is not None:
            hare = hare.next.next
            prev, tort.next, tort = tort, prev, tort.next
        if hare is not None:
            tort = tort.next
        while prev is not None and prev.val == tort.val:
            prev = prev.next
            tort = tort.next
        return prev is None

File: sol/234/234.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct Solution {
    bool isPalindrome(ListNode* head) {
        ListNode *tort = head, *hare = head, *prev = nullptr, *temp;
        while (hare != nullptr && hare->next != nullptr){
            hare = hare->next->next;
            temp = tort;
            tort = temp->next;
            temp->next = prev;
            prev = temp;
        }
        if (hare != nullptr){
            tort = tort->next;
        }
        while (prev != nullptr && prev->val == tort->val){
            prev = prev->next;
            tort = tort->next;
        }
        return prev == nullptr;
    }
};
